Ticket 108: handle calls and state mutation

andioneto opened the pull request:
**Components Added:**
- <Toolbar /> (a toggle button tool bar, basically a radio input that looks like a button)
- <CallControl /> (button group that houses 'hang up' and 'submit form')
- <CallsSideBar /> (parent sidebar component, conditionally renders button groups for call control)
- <PreCallButtonGroup /> (button group for initializing contact call, id bad name, or disconnecting from phonebank)
- <SideBarForm /> (redux form component to connect to the call submission form)

**TL;DR**
- reworked SideBar to better match mock views
- created actions to tie into server endpoints related to basic call page volunteer engagement
- testing for above views (hold on actions to make sure we have the final versions)

**TODO:**
- no action for "call hangup" as the endpoint DNE
- no action for updating status of call on server as endpoint DNE

**To manually test:**
- sign in as user
- join campaign
- start calling
- if you agree to call => call page with active form
- if you choose bad number => will update contact and send new contact to volunteer
- if you choose stop => redirected to home page
- in active call view => select answered => form active and hang up active => hangup => can submit form
- other options just set the status in the store

andioneto committed:
outline for calls sidebar component

lupitadavila committed:
Dotorg 101 call form (#124)

* Added route for CreateContactListContainer to the /admin/contactLists/new path (#121)

Add route to admin sub-app for CreateContactListContainer to the path /admin/contactLists/new

* Bug Fix: Make Description Field Display Properly in Admin Script List View (#123)

Update description field in admin script list view to display the description (was displaying the created_at datetime)

* current progress

andioneto committed:
calls sidebar mock "functioning"

andioneto committed:
Merge branch 'demo' of https://github.com/Twilio-org/phonebank into ao-sidebar-calls

andioneto committed:
added dummy sid to state/store for transitions

andioneto committed:
changed the view to match the requested spec changes

andioneto committed:
working on buttons for sidebar

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-sidebar-calls

andioneto committed:
hang up and submit button mocks are in place

andioneto committed:
actions actions actions

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-sidebar-actions

andioneto committed:
basically done adding actions for the call page sidebar

andioneto committed:
removed call count

andioneto committed:
removed call count stuff

andioneto committed:
updated the users reducer test

andioneto committed:
component testing for calls_sidebar

andioneto committed:
finished all component testing for call sidebar components

andioneto committed:
outline for reducer testing

andioneto committed:
calls reducer test finished

andioneto committed:
made changes to accomodate for pr merging

andioneto committed:
added in hung up

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-call-sidebar-testing

andioneto committed:
fixed merge conflicts, all is well, ready to mergeeeeee

andioneto committed:
fixed linting errors

andioneto committed:
fixed tests

lupitadavila commented on public/src/components/calls/callcntrl_btn_group.jsx:
```diff
@@ -0,0 +1,32 @@
+  );
+};
+
+CallControl.dsiplayName = 'CallControl';
```
did you mean to say displayName? is there a reason why you're setting this?

andioneto replied:
derp, yes

andioneto replied:
✔️

lupitadavila commented on public/src/components/calls/pre_call_btn_grp.jsx:
```diff
@@ -0,0 +1,66 @@
+  handleBadNameClick() {
+    const { call_id, user_id, campaign_id, updateAttempt } = this.props;
+    const outcome = 'BAD_NUMBER';
+    return updateAttempt(user_id, campaign_id, call_id, outcome);
```
I notice there are return statements in some click handlers and not others, could you explain the difference?

andioneto replied:
In this case no, I removed them to be consistent.

lupitadavila commented on public/src/components/calls/side_bar_form.jsx:
```diff
@@ -0,0 +1,29 @@
+  const { handleSubmit } = props;
+  const inputFieldProps = {
+    label: 'notes',
+    name: 'Notes'
```
i think these two can be switched (label is capitalized, name is not)

andioneto replied:
✔️

lupitadavila commented on public/src/components/sub_apps/admin.jsx:
```diff
@@ -19,6 +19,7 @@ import CreateContactListContainer from '../../containers/create_contact_list';
 export default class Admin extends Component {
   componentDidMount() {
     const { id } = this.props.auth;
+    console.log('id in admin compo: ', id);
```
I think this can be removed

andioneto replied:
✔️

andioneto committed:
made some requested changes
made return statements more consistent
fixed Field name/labels
removed a rogue console.log

andioneto committed:
finished writing tests for calls actions
got distracted and cleaned up some of my old tests that had linting errors
consolidated call fixtures

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -1,18 +1,38 @@
+         SET_CALL_CONTACT_INFO,
+         SET_CURRENT_CALL_ACTIVE,
+         SET_CURRENT_CALL_INACTIVE,
+         SET_VOLUNTEER_ACTIVE,
```
Is `SET_VOLUNTEER_CALL_ACTIVE` more explicit. Same goes for the actions below.

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -21,39 +41,91 @@ export function setCurrentCall(callObj) {
+        return new Error('Invalid call status, problem with call assignment.');
+      }
+      if (outcome !== 'PENDING') {
+        return new Error('Invalid outcome, problem with call assignment.');
```
Good assertions!

andioneto replied:
💃

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -21,39 +41,91 @@ export function setCurrentCall(callObj) {
+    }
+  )
+  .then(() => {
+    dispatch(setCurrentCallInactive());
```
I think we can only release a call when we're NOT in an active call i.e. in the pre-call page.
Wouldn't the appropriate dispatch be to clearCurrentCall in this case?

perfectlynormalbeast replied:
And depending on whether this is skip call or end calling, we need to fetch a new call afterwards, right? So perhaps this should function accept another param to decide if it should fetch a new call. Or would you prefer to the calling method to decide it explicitly and call nextCall?

andioneto replied:
Ok just added a check for a call that is in progress.
Also, I changed the .then block to include a conditional dispatch of call assignment or simply clearing the current call.

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -21,39 +41,91 @@ export function setCurrentCall(callObj) {
+  .catch(err => err);
+}
+
+// TODO: waiting for backend to be modified to accept status puts
```
We can still unit test based on API spec and merge this first.

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-call-sidebar-testing

perfectlynormalbeast commented on public/src/components/calls/btn_toolbar.jsx:
```diff
@@ -0,0 +1,43 @@
+                <ToggleButton
+                  bsStyle={style}
+                  value={value.toUpperCase()}
+                  key={value.concat(index)}
```
Why do we need to value+index? Is value not unique in itself?

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/callcntrl_btn_group.jsx:
```diff
@@ -0,0 +1,32 @@
+    <div>
+      <ButtonGroup>
+        <Button
+          disabled={outcome === 'PENDING' && !call_current_active}
```
Is it more explicit to use ```outcome === 'PENDING' && status === 'IN_PROGRESS'```?

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/calls_page_parent.jsx:
```diff
@@ -5,51 +5,85 @@ import CallsPageMain from './calls_page_main';
+            setCurrentCallActive,
+            setCurrentCallInactive,
+            ...storeProps } = this.props;
+    const { id } = current_campaign;
```
Can we not pull out campaign into this ambiguously named id variable? Maybe use campaign_id explicitly?

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -1,13 +1,151 @@
+
   componentDidMount() {
-    console.log(this.props);
+    if (this.props.contact_id) {
```
Since `componentDidMount` and `componentDidUpdate` share the code, maybe break it out into a helper method `refreshSidebarContact`? We'll probably need to bind this to it as well.

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -1,13 +1,151 @@
+  }
+
+  handleStopClick() {
+    const { history } = this.props;
```
We also need to release any call we have with the client right now.

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -1,13 +1,151 @@
+  }
+
+  handleSubmitResponses() {
+    console.log('THIS DOES NOTHING... NOTHING AT ALL!', this.props);
```
I think we can still do the following here:
- hangup the call if it's still in progress
- fetch a new call

andioneto replied:
@nguyenaiden

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -1,13 +1,151 @@
+        />
+      );
+    }
+    if (!!status && (status === 'IN_PROGRESS' || status === 'ATTEMPTED' || status === 'HUNG_UP')) {
```
We should not be seeing these controls for a call that has already been ATTEMPTED i.e. its form has been submitted.

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -1,18 +1,38 @@
   };
 }
 
+export function setCurrentCallActive() {
```
Assuming we decide to keep current_call_active as a property, I think that explicit action functions for setCurrentCallActive and setCurrentCallInactive should not be exported. Other action functions should be causing those store mutations as side effects. Otherwise the client has to "remember" to do those things in addition to calling the methods to place and tearing down those respective calls. For instance, the method that sets call IN_PROGRESS should go ahead and set CURRENT_CALL_ACTIVE = true and likewise for the teardown on HANG_UP and ATTEMPTED.

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -1,13 +1,151 @@
-    console.log(this.props);
+    if (this.props.contact_id) {
+      const { contact_id, getCallContactInfo } = this.props;
+      getCallContactInfo(contact_id);
```
May be a stupid question, but don't we need to dispatch the action objects these methods return? Same applies for all the actions called below?

andioneto replied:
all of these methods passed through props in the container are already mapped to dispatch.

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -21,39 +41,91 @@ export function setCurrentCall(callObj) {
 }
 
-export function clearCallIncrement() {
+export function updateCallStatus(status) {
```
Re: above comment: if we decide to remove the setCurrentCallActive methods, perhaps we can dispatch a SET_CURRENT_CALL_ACTIVE action from here if the status is IN_PROGRESS and SET_CURRENT_CALL_INACTIVE if any other status?

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -1,18 +1,38 @@
+         UPDATE_CALL_STATUS,
+         UPDATE_CALL_OUTCOME,
+         SET_CALL_CONTACT_INFO,
+         SET_CURRENT_CALL_ACTIVE,
```
Is the call active state just an alias for when the call status is IN_PROGRESS? If so, do we really even need to put that in the store as a new variable (and remember to set/clear it when calls are made/hung up?

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -1,13 +1,151 @@
+    const { updateCallOutcome, updateCallStatus } = this.props;
+    if (text === 'ANSWERED') {
+      const { activateCall } = this.props;
+      activateCall();
```
Why are we only activating the call when the call is Answered? Isn't it "active" once it's been dialled by putting it IN_PROGRESS?

andioneto replied:
active removed, ✔️

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -1,13 +1,151 @@
+
+  handleHangUp() {
+    const { inactivateCall, updateCallStatus } = this.props;
+    inactivateCall();
```
Yeah, if we're activating the call when ANSWERED and inactivating the call when HUNG_UP what purpose does it actually serve? Also, it's weird to couple the outcome (ANSWERED) with the status (HUNG_UP) meaning that a outcome transition makes it active and a status transition makes it inactive.

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/pre_call_btn_grp.jsx:
```diff
@@ -0,0 +1,66 @@
+
+  handleStartCallingClick() {
+    const { updateCallStatus } = this.props;
+    updateCallStatus('IN_PROGRESS');
```
Same comment as before - do we need to dispatch these actions? Or should the action functions themselves take care of the dispatching? We should be consistent in that. Either the action takes that responsibility or the component does. I favor action method doing so.

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/pre_call_btn_grp.jsx:
```diff
@@ -0,0 +1,66 @@
+
+  handleSkipClick() {
+    const { call_id, user_id, campaign_id, releaseCall, nextCall } = this.props;
+    releaseCall(user_id, campaign_id, call_id);
```
re: previous comment about releasing calls - would it make sense for this method to take another param, maybe `assign_new_call=true|false` to see if it should internally invoke nextCall after the call has been released?

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/components/calls/pre_call_btn_grp.jsx:
```diff
@@ -0,0 +1,66 @@
+  handleBadNameClick() {
+    const { call_id, user_id, campaign_id, updateAttempt } = this.props;
+    const outcome = 'BAD_NUMBER';
+    updateAttempt(user_id, campaign_id, call_id, outcome);
```
This should also go ahead and assign us a new call?

andioneto replied:
Yep, that happens in the .then block of updateCallAttempt()

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -1,18 +1,38 @@
 }
-export function promoteNextToCurrent() {
+
+export function clearVolunteerActive() {
```
This method should also DELETE to users/:id/camp/:id/calls to terminate the volunteer call?

andioneto replied:
That's in the ticket I just did.

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -1,18 +1,38 @@
+         CLEAR_VOLUNTEER_ACTIVE } from '../reducers/calls';
 
-export function incrementCallCount() {
+export function setVolunteerActive() {
```
This method should also POST to users/:id/camp/:id/calls to set up the volunteer call?

perfectlynormalbeast commented on public/src/components/common/button_objs.jsx:
```diff
@@ -112,6 +112,7 @@ export default {
         size: 'xsmall',
         style: 'success',
         handler: () => {
+          // this will need to set a dummy user_sid in the state
```
Um no it doesn't. Joining a campaign should not set any call SID. Also we decided against putting call SIDs in the client store.

andioneto replied:
✔️

perfectlynormalbeast reviewed (changes requested):
Sorry but got a bunch of comments.
Also I'll take a look at tests once we've addressed them.

andioneto committed:
review remarks resolved
ready for PR review

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -103,14 +89,20 @@ export function assignToCall(userId, campaignId) {
 
-export function releaseCall(userId, campaignId, callId) {
+export function releaseCall(userId, campaignId, callId, currentCallStatus, next = false) {
+  if (currentCallStatus === 'IN_PROGRESS') {
```
Actually let's be even more careful - `if (currentCallStatus !== 'ASSIGNED')`?

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -103,14 +89,20 @@ export function assignToCall(userId, campaignId) {
   .then(() => {
-    dispatch(setCurrentCallInactive());
+    if (next) {
+      return dispatch(assignToCall(userId, campaignId));
```
Woah - here we're assigning a call and then clearing it in the next line. No bueno.

perfectlynormalbeast commented on public/src/components/calls/calls_sidebar.jsx:
```diff
@@ -35,13 +35,10 @@ export default class CallsSideBar extends Component {
-      activateCall();
-      return;
+    if (text !== 'ANSWERED') {
+      updateCallStatus('HUNG_UP');
```
This is a bad idea because now all of our outcome buttons are also hang up buttons. :( Let's say that only the Hang Up button triggers that action. I think this click should just update outcome and nothing else.

perfectlynormalbeast commented on server/config/auth/jwt.js:
```diff
@@ -1,8 +1,11 @@
 import passport from 'passport';
 import jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
```
Let's pull out dotenv and Twilio things.

andioneto replied:
✔️

perfectlynormalbeast commented on public/src/store.jsx:
```diff
@@ -38,9 +38,9 @@ const middleware = [immutable(), createLogger(), promise(), thunk];
 
 // to reset the store:
-// persistStore(store).purge();
+persistStore(store).purge();
```
This was off the whole time? 😮

andioneto replied:
Nope, just for updating the reducers

perfectlynormalbeast reviewed (changes requested):
moar comments

andioneto committed:
removed dotenv relics

andioneto committed:
turned back on persist store

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-call-sidebar-testing

andioneto committed:
deleted the twilio.js file bc it was contaminated, aded master version back

andioneto committed:
Merge branch 'master' into ao-call-sidebar-testing

perfectlynormalbeast commented on public/src/components/campaign/sidebar_list_view/current_campaign.jsx:
```diff
@@ -17,6 +12,7 @@ const CurrentCampaign = (props) => {
   }
   const handleClick = () => {
     const { id: campaignId } = id;
+    setVolunteerActive();
```
Isn't this only set when the volunteer starts calling by clicking Start Calling on the calls page? Why are we triggering this when they're just going to the calls page?

andioneto replied:
The start calling button is on this page, then this takes the volunteer to the control view (after setting connection to active) and the call pre-call view (in component did mount) assigns a call to the user.

perfectlynormalbeast reviewed (changes requested):
Left one more comment. Also there are 3 comments from my last review that you didn't get to. Let me know if you need me to repost them or if you can just access them from the out of date diffs.

andioneto committed:
removed comment

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-call-sidebar-testing

andioneto committed:
Merge branch 'ao-call-sidebar-testing' of https://github.com/Twilio-org/phonebank into ao-call-sidebar-testing

andioneto committed:
error handling in releaseCall action: changed error condition to throw if status is !== assigned

andioneto committed:
fixed testing and ready for review again

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-call-sidebar-testing

perfectlynormalbeast commented on public/src/components/calls/callcntrl_btn_group.jsx:
```diff
@@ -1,29 +1,21 @@
-  // TODO: SUBMIT DOES NOTHING RN... NOTHING
-  const { handleHangUp, submitHandler, outcome, status } = props;
+  const { handler, text, outcome, status, htmlID } = props;
+  const style = text === 'Hang Up' ? 'danger' : 'success';
```
One could argue that we should do something similar to buttons where the component just accepts props for everything. But you know what. This PR has had enough review. We can let this slide. Progress over perfection.
